config/configoptional/optional_test.go

```diff
@@ -365,3 +367,94 @@ func TestComparePointer(t *testing.T) {
 		})
 	}
 }
+
+func TestOptionalMarshal(t *testing.T) {
+	tests := []struct {
+		name     string
+		value    Config[Sub]
+		expected map[string]any
+	}{
+		{
+			name:     "none (zero value)",
+			value:    Config[Sub]{},
+			expected: map[string]any{"sub": nil},
+		},
+		{
+			name:     "none",
+			value:    Config[Sub]{Sub1: None[Sub]()},
+			expected: map[string]any{"sub": nil},
+		},
+		{
+			name:     "default",
+			value:    Config[Sub]{Sub1: Default(subDefault)},
+			expected: map[string]any{"sub": nil},
+		},
+		{
+			name: "some",
+			value: Config[Sub]{Sub1: Some(Sub{
+				Foo: "bar",
+			})},
+			expected: map[string]any{
+				"sub": map[string]any{
+					"foo": "bar",
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			conf := confmap.New()
+			require.NoError(t, conf.Marshal(test.value))
+			assert.Equal(t, test.expected, conf.ToStringMap())
+		})
+	}
+}
+
+func TestComparePointerMarshal(t *testing.T) {
+	type Wrap[T any] struct {
+		Sub1 T `mapstructure:"sub"`
```

Comment:
```suggestion
		// Note: passes without requiring "squash".
		Sub1 T `mapstructure:"sub"`
```
I noticed that adding `squash` causes the cases in this test to pass without any of the confmap changes. In other words, the changes in confmap are what allow us to leave out the tag and still have equivalent functionality to pointers.

Let me know if you think this comment misses the point or is superfluous, but I think it's probably worth calling out.